htmresearch/algorithms/hierarchical_clustering.py

```diff
@@ -0,0 +1,113 @@
+import numpy
+import scipy.sparse
+import scipy.cluster.hierarchy
+import matplotlib.pyplot as plt
+from nupic.algorithms.KNNClassifier import KNNClassifier
+
+
+
+class HierarchicalClustering(object):
+  """
+  Implements hierarchical agglomerative clustering on the output of a classification network.
+  """
+
+
+  def __init__(self, classificationNetwork, linkageMethod='single'):
+    knn = classificationNetwork.getKNN()
+    dim = len(knn.getPattern(0, sparseBinaryForm=False))
+    sparseRowList = []
+
+    for i in xrange(knn._numPatterns):
+      nzIndices = knn.getPattern(i, sparseBinaryForm=True)
+      sparseRow = scipy.sparse.csr_matrix(
+        (numpy.ones(len(nzIndices), dtype=bool),
+        (numpy.zeros(len(nzIndices)), nzIndices)),
+        shape=(1,dim))
+      sparseRowList.append(sparseRow)
+
+    sparseDataMatrix = scipy.sparse.vstack(sparseRowList)
+    overlaps = HierarchicalClustering._computeOverlaps(sparseDataMatrix)
```

Comment:
This and the subsequent call to cluster could take a while. Typically constructors should return fast. Perhaps put the heavy weight code in a separate method that is explicitly called?